### PR TITLE
Research terminal tgui

### PR DESCRIPTION
**Pull request by mullenpaul**
## About The Pull Request

I have ported the research terminal and the published medhud from nanoui to tgui. I have also created a CRT theme which can be used for new terminals to improve the vibes.

### Research Management Panel

image

Purchase document confirmation message now in-game rather than popup

image

CL message to improve level now in-game message rather than popup

image

### View Chemical Panel

image

You can hide old scans, only displaying most recent

image

### Published List (Terminal)

image

### Published HUD

image

### Green CRT

image

### Blue CRT

image

### Manage Publish from CMO terminal and see error

image

## Why It's Good For The Game

Migration from nanoui to TGUI needs to happen.

CRT theme gives me flashbacks to using MSDOS in the 90s.

## Changelog

:cl:
refactor: Research terminal now using TGUI
refactor: Medhud published papers now using TGUI
refactor: Added CRT theme in yellow, green and blue
refactor: Reworked research chemicals documents to allow for metadata to be stored
refactor: Improved sorting options for chemicals
refactor: Research paper screen auto-hides olds scans, displaying only the most recent scan. This can be toggled.
refactor: Moved most of the ui_chat messages regarding terminal errors to appear as UI elements.
/:cl:

### CODE REVIEW

**Comment by mullenpaul** [comment]
It would be great to get feedback on the SCSS - I don't know the framework very well so feedback there would be greatly appreciated.

**Review by stanalbatross — CHANGES_REQUESTED** [review]
the only thing I can immediately see wrong is no ui_state for the terminal or glasses

**Comment by mullenpaul** [comment]
Resolved ui_state for research terminal and health hud

File: code/modules/reagents/chemical_research/Chemical-Research.dm [hunk 1]
```diff
@@ -24,21 +24,59 @@ var/global/datum/chemical_data/chemical_data = new /datum/chemical_data/
...
+		"document"=R
+	)
+	research_documents["[document_type]"] += list(new_document)
+
+/datum/chemical_data/proc/get_report(var/doc_type, var/doc_title)
+	var/obj/item/paper/research_report/report = null
+	for(var/i in chemical_data.research_documents[doc_type])
+		if(i["document_title"] == doc_title)
+			report = i["document"]
+			break
```
Comment (Watermelon914) [inline code/modules/reagents/chemical_research/Chemical-Research.dm:41]: use a detailed variable name for i here.

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/modules/reagents/chemical_research/Chemical-Research.dm (hunk 1 above)
Comment (mullenpaul) [reply code/modules/reagents/chemical_research/Chemical-Research.dm]: Resolved this.

**Review by harryob — CHANGES_REQUESTED** [review]
handling for property papers without analysed properties - can't see anything else a matter

File: tgui/packages/tgui/interfaces/ResearchTerminal.tsx [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+          </Stack.Item>
+        </Stack>
+      </Stack.Item>
+    </Stack>);
+};
+
+const NoCompoundsDetected = (_, context) => {
+  return (
+    <span>
+      Error: no chemicals have been detected.
```
[the remark this answers is not in the document]
Comment (mullenpaul) [reply tgui/packages/tgui/interfaces/ResearchTerminal.tsx]: Done

**Review by harryob — APPROVED** [review]

**Comment by CocaColaTastesGood** [comment]
Passing a float as the tier when purchasing a document might default it to tier 5 despite costing practically nothing. Not sure how it works in practice though.

File: tgui/packages/tgui/interfaces/ResearchTerminal.tsx [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+      {clearance_level < 5
+        && (
+          <Button
+            disabled={isDisabled}
+            onClick={() => {
+              setSelectedTab(1);
+              setConfirm("broker_clearance");
+            }}
+          >
+            Improve {data.broker_cost}
```
Comment (harryob) [inline tgui/packages/tgui/interfaces/ResearchTerminal.tsx:545]:
```suggestion
            Improve {data.broker_cost}CR
```

File: tgui/packages/tgui/interfaces/ResearchTerminal.tsx [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+              setConfirm(undefined);
+            }}
+            onCancel={() => setConfirm(undefined)}
+          >
+            <span>
+              The CL can swipe their ID card on the console to increase
+              clearance for free, given enough DEFCON. Are you sure you want to
+              spend <u>{data.broker_cost}</u> research credits to increase the
+              clearance immediately?
+            </span>
```
Comment (harryob) [inline tgui/packages/tgui/interfaces/ResearchTerminal.tsx:393]:
```suggestion
            <span>
              Are you sure you want to spend <u>{data.broker_cost}</u> research 
              credits to increase the clearance immediately?
            </span>
```
the ship CL doesn't have this ability regardless of intel level - and is a surv CL thing. this message could be made to reference the ability of the survivor CL to increase clearance, but i'm not sure how you do that neatly

**Review by harryob — CHANGES_REQUESTED** [review]
looks all good - just some nitpicks

File: tgui/packages/tgui/interfaces/ResearchTerminal.tsx (hunk 4 above)
Comment (mullenpaul) [reply tgui/packages/tgui/interfaces/ResearchTerminal.tsx:393]: I just copied the existing text, probably just best to ignore survivor cl info for rp reasons

**Comment by mullenpaul** [comment]
Well I screwed up my git on here.

**Pull request closed without merging**